module4-authentication-and-security/r1.1.1-authentication-persistence/README.md

```diff
@@ -0,0 +1,268 @@
+
+In Express.js, we can use the library [express-session](https://www.npmjs.com/package/express-session) to create session storage. It can be plugged into the Express app as a middleware that attaches a `session` object to any request that comes to the server.
+
+```js
+// ./app.js
+
+const express = require("express");
+const sessions = require("express-session");
+
+// Init our express app
+const app = express();
+
+// Below is where we introduce encrypted sessions
+// Make sure your app secret is unique and
+// defined by cryptographic random generator.
+// It has in-built ability to parse and use cookies
+app.use(
+  sessions({
+    secret: "MY_APP_SECRET",
+    cookie: { secure: false }, // make sure to change this to true on production code
+  })
+);
+
+let listener = app.listen(5000, function () {
+  console.log("Listening on port " + listener.address().port);
+});
+```
+
+In the code snippet above, we configured the session storage using `app.use()` and created a new session instance with the configuration below:
+
+- `secret`: a token that will be used to encrypt the data and cookies. It can also be a universal `APP_SECRET` defined in the environment variables. This must be a random secret generated by secure random generators.
+- `cookie`: cookie parameters that will be set in the client. Because our development environment doesn't have HTTPs, we turned off `secure`. In a production environment, the cookie should only be sent in secure contexts so secure should be true.
+
+This initialization would eventually activate the `session` object in each request, and a `Set-Cookie` header to each response when a session is created. The cookie will contain the `sessionID` that the server can use to read the session storage and grab the session data.
+
+When a cookie is set in the client or browser, the client will always send that cookie with subsequent requests to the original server that issued the cookie. The cookie will be sent as part of the request headers. That's why it is important to ensure it is only sent in secure contexts, so it is encrypted and an eavesdropper can't [hijack the session](https://en.wikipedia.org/wiki/Session_hijacking).
+
+![Session Diagram](../assets/use-of-session-cookies.webp)
+
+This can be used to persist authentication as well. This is often done when a user logs in and activates the **Remember Me** function. When login validation is carried out successfully, the server should also save a user copy to the session as seen in the previous lesson:
+
+```js
+req.session.user = user;
+```
+
+And when requesting protected resources, we can check if the request session that is tied to the request cookies, has the user or not. If it has the user as stored in our login logic, it means the request is authentic.
```

Comment:
don't really understand the section after line 53, there's no context for this code 

also why can't someone just spoof the session user

Reply:
The context would come from lesson 1.1 as this is a descendent of that. 

And makes sense for spoofing, which is also tied to cookies security and https. I think it would make sense to talk about this when we cover https, as one could spoof jwt as well. 

I am considering introducing cookies security in this section too